core/src/messages/agent.rs

```diff
@@ -241,6 +241,37 @@ pub struct SpawnRequest {
 
     /// Credentials used to fetch the image.
     pub credentials: Option<DockerCredentials>,
+
+    /// Resource limits
+    pub resource_limits: Option<ResourceLimits>,
+}
+
+// eventually, this will be generic over executors
+// currently only applies to docker
+#[serde_as]
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct ResourceLimits {
+    /// period of cpu time, serializes as microseconds
+    #[serde_as(as = "DurationMicroSeconds")]
+    pub cpu_period: Duration,
+
+    /// proportion of period used by container
+    pub cpu_period_percent: u8,
+
+    /// total cpu time allocated to container    
+
+    #[serde_as(as = "DurationSeconds")]
+    pub cpu_time_limit: Duration,
+}
+
+impl Default for ResourceLimits {
+    fn default() -> ResourceLimits {
+        ResourceLimits {
+            cpu_period: Duration::from_millis(100),
+            cpu_period_percent: 100,
+            cpu_time_limit: Duration::from_secs(30 * 60),
```

Comment:
I agree with this. Spawner shouldn't be opinionated about its workloads; if these limits make sense in the context of how we employ Spawner we should set them in p2.

Reply:
I guess this is a general instance of “move optionals to the leaves of the data structure” which I think is generally a good idea. The exception is where the sub-data-structure being missing represents different semantics to when the leaves are missing.

E.g. if we have a data structure

```
struct Foobar {
    foo: Option<String>,
    bar: Option<String>,
}
```

We should only use an `Option<Foobar>` if a value of `None` conveys different semantics to `Foobar {foo: None, bar: None}`. Otherwise, we have multiple ways of representing the same thing, which is not ideal.